featuretools/feature_discovery/feature_discovery.py

```diff
@@ -0,0 +1,413 @@
+from itertools import combinations, permutations, product
+from typing import Dict, List, Type
+
+import woodwork.type_sys.type_system as ww_type_system
+from woodwork.column_schema import ColumnSchema
+from woodwork.logical_types import LogicalType
+from woodwork.table_schema import TableSchema
+
+from featuretools.feature_discovery.type_defs import ANY, Feature
+from featuretools.primitives.base.primitive_base import PrimitiveBase
+
+
+def column_to_keys(column_schema: ColumnSchema) -> List[str]:
+    """
+    Generate hashing keys from Columns Schema. For example:
+    - ColumnSchema(logical_type=Double) -> ["Double"]
+    - ColumnSchema(semantic_tags={"index"}) -> ["index"]
+    - ColumnSchema(logical_type=Double, semantic_tags={"index", "other"}) -> ["Double,index", "Double,other"]
+
+    Args:
+        column_schema (ColumnSchema):
+
+    Returns:
+        List[str]
+            List of hashing keys
+    """
+    keys: List[str] = []
+    lt = column_schema.logical_type
+    tags = column_schema.semantic_tags
+    lt_key = None
+    if lt:
+        lt_key = type(lt).__name__
+
+    if len(tags) > 0:
+        for tag in tags:
+            if tag is not None:
+                keys.append(f"{lt_key},{tag}" if lt_key is not None else tag)
+
+    elif lt_key is not None:
+        keys.append(lt_key)
+    else:
+        keys.append(ANY)
+
+    return keys
+
+
+def index_input_set(input_set: List[ColumnSchema]) -> Dict[str, int]:
+    """
+    Indexes input set to find types of columns and the quantity of eatch
+
+    Args:
+        input_set (List(ColumnSchema)):
+            List of Column types needed by associated primitive.
+
+    Returns:
+        Dict[str, int]
+            A hashmap from key to int
+
+    Examples:
+        .. code-block:: python
+
+            from featuretools.feature_discovery.feature_discovery import get_features
+            from woodwork.column_schema import ColumnSchema
+
+            input_set = [ColumnSchema(semantic_tags={"numeric"}), ColumnSchema(semantic_tags={"numeric"})]
+            indexed_input_set = index_input_set(input_set)
+            {"numeric": 2}
+    """
+    out = {}
+    for column_schema in input_set:
+        for key in column_to_keys(column_schema):
+            out[key] = out.get(key, 0) + 1
+    return out
+
+
+def get_features(
+    feature_groups: Dict[str, List[Feature]],
+    input_set: List[ColumnSchema],
```

Comment:
Might be missing something, but can we call this 'required_columns_set` or something? (Unless there is another meaning of `input_set` I am missing.)